{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d867221d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "Warning message:\n",
      "\"package 'matrixStats' was built under R version 4.3.2\"\n",
      "\n",
      "Attaching package: 'MatrixGenerics'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:matrixStats':\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: 'BiocGenerics'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:stats':\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: 'S4Vectors'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:utils':\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "\n",
      "Attaching package: 'IRanges'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:grDevices':\n",
      "\n",
      "    windows\n",
      "\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'Biobase'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:MatrixGenerics':\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:matrixStats':\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "Loading required package: usethis\n",
      "\n",
      "Loading required package: Seurat\n",
      "\n",
      "Attaching SeuratObject\n",
      "\n",
      "Seurat v4 was just loaded with SeuratObject v5; disabling v5 assays and\n",
      "validation routines, and ensuring assays work in strict v3/v4\n",
      "compatibility mode\n",
      "\n",
      "\n",
      "Attaching package: 'Seurat'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:SummarizedExperiment':\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "Loading required package: dplyr\n",
      "\n",
      "Warning message:\n",
      "\"package 'dplyr' was built under R version 4.3.2\"\n",
      "\n",
      "Attaching package: 'dplyr'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:Biobase':\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:GenomicRanges':\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from 'package:GenomeInfoDb':\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:IRanges':\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:S4Vectors':\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:BiocGenerics':\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from 'package:matrixStats':\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:stats':\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Loading required package: HGNChelper\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(SingleCellExperiment)\n",
    "library(devtools)\n",
    "library(scAnnoX)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "be308963",
   "metadata": {},
   "outputs": [],
   "source": [
    "# A preprocessed dataset of type Seurat.\n",
    "test.obj <- readRDS('../data/GSE81252.test.obj')\n",
    "ref.obj <- readRDS('../data/GSE81252.ref.obj')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7dbf9292",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "b3b120ed",
   "metadata": {},
   "source": [
    "### Only one single cell annotation tool is invoked"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3d789ac1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 10 × 8</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>orig.ident</th><th scope=col>nCount_RNA</th><th scope=col>nFeature_RNA</th><th scope=col>cell_id</th><th scope=col>experiment</th><th scope=col>assignment_LB</th><th scope=col>CellType</th><th scope=col>SingleR</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>A1_he2</th><td>A1 </td><td>29630.19</td><td>7229</td><td>A1_he2   </td><td>he2  </td><td>HE   </td><td>HE   </td><td>HE   </td></tr>\n",
       "\t<tr><th scope=row>A1_huvec</th><td>A1 </td><td>30603.64</td><td>7374</td><td>A1_huvec </td><td>huvec</td><td>EC   </td><td>EC   </td><td>EC   </td></tr>\n",
       "\t<tr><th scope=row>A10_he1</th><td>A10</td><td>26921.42</td><td>6185</td><td>A10_he1  </td><td>he1  </td><td>HE   </td><td>HE   </td><td>HE   </td></tr>\n",
       "\t<tr><th scope=row>A10_lb2</th><td>A10</td><td>21383.21</td><td>4296</td><td>A10_lb2  </td><td>lb2  </td><td>EC-LB</td><td>EC-LB</td><td>EC-LB</td></tr>\n",
       "\t<tr><th scope=row>A10_lb5</th><td>A10</td><td>22637.05</td><td>5307</td><td>A10_lb5  </td><td>lb5  </td><td>MC-LB</td><td>MC-LB</td><td>MC-LB</td></tr>\n",
       "\t<tr><th scope=row>A11_he1</th><td>A11</td><td>16551.88</td><td>3725</td><td>A11_he1  </td><td>he1  </td><td>HE   </td><td>HE   </td><td>HE-LB</td></tr>\n",
       "\t<tr><th scope=row>A11_lb5</th><td>A11</td><td>24529.42</td><td>5102</td><td>A11_lb5  </td><td>lb5  </td><td>EC-LB</td><td>EC-LB</td><td>EC-LB</td></tr>\n",
       "\t<tr><th scope=row>A12_he1</th><td>A12</td><td>24989.38</td><td>5695</td><td>A12_he1  </td><td>he1  </td><td>HE   </td><td>HE   </td><td>HE   </td></tr>\n",
       "\t<tr><th scope=row>A12_huvec</th><td>A12</td><td>32676.21</td><td>7801</td><td>A12_huvec</td><td>huvec</td><td>EC   </td><td>EC   </td><td>EC   </td></tr>\n",
       "\t<tr><th scope=row>A12_msc2</th><td>A12</td><td>23993.08</td><td>6268</td><td>A12_msc2 </td><td>msc2 </td><td>MC   </td><td>MC   </td><td>MC   </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 10 × 8\n",
       "\\begin{tabular}{r|llllllll}\n",
       "  & orig.ident & nCount\\_RNA & nFeature\\_RNA & cell\\_id & experiment & assignment\\_LB & CellType & SingleR\\\\\n",
       "  & <fct> & <dbl> & <int> & <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\tA1\\_he2 & A1  & 29630.19 & 7229 & A1\\_he2    & he2   & HE    & HE    & HE   \\\\\n",
       "\tA1\\_huvec & A1  & 30603.64 & 7374 & A1\\_huvec  & huvec & EC    & EC    & EC   \\\\\n",
       "\tA10\\_he1 & A10 & 26921.42 & 6185 & A10\\_he1   & he1   & HE    & HE    & HE   \\\\\n",
       "\tA10\\_lb2 & A10 & 21383.21 & 4296 & A10\\_lb2   & lb2   & EC-LB & EC-LB & EC-LB\\\\\n",
       "\tA10\\_lb5 & A10 & 22637.05 & 5307 & A10\\_lb5   & lb5   & MC-LB & MC-LB & MC-LB\\\\\n",
       "\tA11\\_he1 & A11 & 16551.88 & 3725 & A11\\_he1   & he1   & HE    & HE    & HE-LB\\\\\n",
       "\tA11\\_lb5 & A11 & 24529.42 & 5102 & A11\\_lb5   & lb5   & EC-LB & EC-LB & EC-LB\\\\\n",
       "\tA12\\_he1 & A12 & 24989.38 & 5695 & A12\\_he1   & he1   & HE    & HE    & HE   \\\\\n",
       "\tA12\\_huvec & A12 & 32676.21 & 7801 & A12\\_huvec & huvec & EC    & EC    & EC   \\\\\n",
       "\tA12\\_msc2 & A12 & 23993.08 & 6268 & A12\\_msc2  & msc2  & MC    & MC    & MC   \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 10 × 8\n",
       "\n",
       "| <!--/--> | orig.ident &lt;fct&gt; | nCount_RNA &lt;dbl&gt; | nFeature_RNA &lt;int&gt; | cell_id &lt;chr&gt; | experiment &lt;chr&gt; | assignment_LB &lt;chr&gt; | CellType &lt;chr&gt; | SingleR &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| A1_he2 | A1  | 29630.19 | 7229 | A1_he2    | he2   | HE    | HE    | HE    |\n",
       "| A1_huvec | A1  | 30603.64 | 7374 | A1_huvec  | huvec | EC    | EC    | EC    |\n",
       "| A10_he1 | A10 | 26921.42 | 6185 | A10_he1   | he1   | HE    | HE    | HE    |\n",
       "| A10_lb2 | A10 | 21383.21 | 4296 | A10_lb2   | lb2   | EC-LB | EC-LB | EC-LB |\n",
       "| A10_lb5 | A10 | 22637.05 | 5307 | A10_lb5   | lb5   | MC-LB | MC-LB | MC-LB |\n",
       "| A11_he1 | A11 | 16551.88 | 3725 | A11_he1   | he1   | HE    | HE    | HE-LB |\n",
       "| A11_lb5 | A11 | 24529.42 | 5102 | A11_lb5   | lb5   | EC-LB | EC-LB | EC-LB |\n",
       "| A12_he1 | A12 | 24989.38 | 5695 | A12_he1   | he1   | HE    | HE    | HE    |\n",
       "| A12_huvec | A12 | 32676.21 | 7801 | A12_huvec | huvec | EC    | EC    | EC    |\n",
       "| A12_msc2 | A12 | 23993.08 | 6268 | A12_msc2  | msc2  | MC    | MC    | MC    |\n",
       "\n"
      ],
      "text/plain": [
       "          orig.ident nCount_RNA nFeature_RNA cell_id   experiment assignment_LB\n",
       "A1_he2    A1         29630.19   7229         A1_he2    he2        HE           \n",
       "A1_huvec  A1         30603.64   7374         A1_huvec  huvec      EC           \n",
       "A10_he1   A10        26921.42   6185         A10_he1   he1        HE           \n",
       "A10_lb2   A10        21383.21   4296         A10_lb2   lb2        EC-LB        \n",
       "A10_lb5   A10        22637.05   5307         A10_lb5   lb5        MC-LB        \n",
       "A11_he1   A11        16551.88   3725         A11_he1   he1        HE           \n",
       "A11_lb5   A11        24529.42   5102         A11_lb5   lb5        EC-LB        \n",
       "A12_he1   A12        24989.38   5695         A12_he1   he1        HE           \n",
       "A12_huvec A12        32676.21   7801         A12_huvec huvec      EC           \n",
       "A12_msc2  A12        23993.08   6268         A12_msc2  msc2       MC           \n",
       "          CellType SingleR\n",
       "A1_he2    HE       HE     \n",
       "A1_huvec  EC       EC     \n",
       "A10_he1   HE       HE     \n",
       "A10_lb2   EC-LB    EC-LB  \n",
       "A10_lb5   MC-LB    MC-LB  \n",
       "A11_he1   HE       HE-LB  \n",
       "A11_lb5   EC-LB    EC-LB  \n",
       "A12_he1   HE       HE     \n",
       "A12_huvec EC       EC     \n",
       "A12_msc2  MC       MC     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Using SingleR as an example.\n",
    "pred.obj <- autoAnnoTools(\n",
    "    test.obj,\n",
    "    ref.obj = ref.obj,\n",
    "    ref.ctype = 'CellType',\n",
    "    marker.lst = marker.lst,\n",
    "    method = 'SingleR',\n",
    "    select.marker = 'Seurat',\n",
    "    top.k = 30,\n",
    "    strategy = 'refernce-based'\n",
    ") \n",
    "head(pred.obj)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8258952d",
   "metadata": {},
   "outputs": [],
   "source": [
    "?findMarkerToolsForSc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2143a1d2",
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in findMarkerToolsForSc(ref.obj, to.list = TRUE, top.k = 30): could not find function \"findMarkerToolsForSc\"\n",
     "output_type": "error",
     "traceback": [
      "Error in findMarkerToolsForSc(ref.obj, to.list = TRUE, top.k = 30): could not find function \"findMarkerToolsForSc\"\nTraceback:\n"
     ]
    }
   ],
   "source": [
    "# Using SCINA as an example.\n",
    "Idents(ref.obj) <- ref.obj$CellType\n",
    "marker.lst <- findMarkerToolsForSc(ref.obj, to.list = TRUE, top.k = 30)\n",
    "pred.obj <- autoAnnoTools(\n",
    "    test.obj,\n",
    "    ref.obj = ref.obj,\n",
    "    ref.ctype = 'CellType',\n",
    "    marker.lst = marker.lst,\n",
    "    method = 'SCINA',\n",
    "    select.marker = 'Seurat',\n",
    "    top.k = 30,\n",
    "    strategy = 'marker-based'\n",
    ") \n",
    "head(pred.obj)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac38cd2a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"The following arguments are not used: drop\"\n"
     ]
    }
   ],
   "source": [
    "a <- autoAnnoResult(pred.obj)\n",
    "a"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08414153",
   "metadata": {},
   "source": [
    "### Multiple single-cell annotation tools are invoked"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "26b5c235",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Loading \u001b[34mscAnnoX\u001b[39m\n",
      "Warning message:\n",
      "\"replacing previous import 'data.table::shift' by 'tictoc::shift' when loading 'CelliD'\"\n"
     ]
    }
   ],
   "source": [
    "devtools::load_all('../../scAnnoX')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b18e3c11",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Updating \u001b[34mscAnnoX\u001b[39m documentation\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Loading \u001b[34mscAnnoX\u001b[39m\n"
     ]
    }
   ],
   "source": [
    "document()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "76c83272",
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in eval(expr, envir, enclos): object 'e' not found\n",
     "output_type": "error",
     "traceback": [
      "Error in eval(expr, envir, enclos): object 'e' not found\nTraceback:\n",
      "1. findMarkerToolsForSc(e)",
      "2. findMarkersBySeurat(obj.seu, to.list, top.k, ...)   # at line 103-106 of file C:/Users/13123/Desktop/scAnnoX-main/R/findMarkerTools.R",
      "3. FindAllMarkers(obj.seu, ...)   # at line 126 of file C:/Users/13123/Desktop/scAnnoX-main/R/findMarkerTools.R",
      "4. sort(x = unique(x = Idents(object = object)))",
      "5. unique(x = Idents(object = object))",
      "6. Idents(object = object)"
     ]
    }
   ],
   "source": [
    "findMarkerToolsForSc(e)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f70d4d86",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<dl>\n",
       "\t<dt>$SingleR</dt>\n",
       "\t\t<dd><pre class=language-r><code>function (obj.seu, ref.obj, ref.ctype = NULL, ...) \n",
       "{\n",
       "<span style=white-space:pre-wrap>    test.expr &lt;- GetAssayData(obj.seu, slot = \"data\") %&gt;% as.matrix</span>\n",
       "<span style=white-space:pre-wrap>    ref.expr &lt;- GetAssayData(ref.obj, slot = \"data\") %&gt;% as.matrix</span>\n",
       "<span style=white-space:pre-wrap>    if (!is.null(ref.ctype)) </span>\n",
       "<span style=white-space:pre-wrap>        label.ref &lt;- ref.obj@meta.data[, ref.ctype]</span>\n",
       "<span style=white-space:pre-wrap>    else label.ref &lt;- Idents(ref.obj)</span>\n",
       "<span style=white-space:pre-wrap>    res.pred &lt;- SingleR::SingleR(test = test.expr, ref = ref.expr, </span>\n",
       "<span style=white-space:pre-wrap>        labels = label.ref, ...)</span>\n",
       "<span style=white-space:pre-wrap>    obj.seu$SingleR &lt;- res.pred$labels</span>\n",
       "<span style=white-space:pre-wrap>    return(obj.seu)</span>\n",
       "}</code></pre></dd>\n",
       "\t<dt>$Seurat</dt>\n",
       "\t\t<dd><pre class=language-r><code>function (obj.seu, ref.obj, ref.ctype = NULL, npcs = 30, ...) \n",
       "{\n",
       "<span style=white-space:pre-wrap>    anchors &lt;- FindTransferAnchors(reference = ref.obj, query = obj.seu, </span>\n",
       "<span style=white-space:pre-wrap>        dims = 1:npcs)</span>\n",
       "<span style=white-space:pre-wrap>    if (!is.null(ref.ctype)) </span>\n",
       "<span style=white-space:pre-wrap>        label.ref &lt;- ref.obj@meta.data[, ref.ctype]</span>\n",
       "<span style=white-space:pre-wrap>    else label.ref &lt;- Idents(ref.obj)</span>\n",
       "<span style=white-space:pre-wrap>    res.pred &lt;- TransferData(anchorset = anchors, refdata = label.ref, </span>\n",
       "<span style=white-space:pre-wrap>        dims = 1:npcs, ...)</span>\n",
       "<span style=white-space:pre-wrap>    obj.seu$SeuratAnno &lt;- res.pred[, 1]</span>\n",
       "<span style=white-space:pre-wrap>    return(obj.seu)</span>\n",
       "}</code></pre></dd>\n",
       "\t<dt>$sciBet</dt>\n",
       "\t\t<dd><pre class=language-r><code>function (obj.seu, ref.obj, ref.ctype = NULL, ...) \n",
       "{\n",
       "<span style=white-space:pre-wrap>    test.df &lt;- GetAssayData(obj.seu, slot = \"data\") %&gt;% as.data.frame %&gt;% </span>\n",
       "<span style=white-space:pre-wrap>        t</span>\n",
       "<span style=white-space:pre-wrap>    ref.df &lt;- GetAssayData(ref.obj, slot = \"data\") %&gt;% as.data.frame %&gt;% </span>\n",
       "<span style=white-space:pre-wrap>        t</span>\n",
       "<span style=white-space:pre-wrap>    if (!is.null(ref.ctype)) </span>\n",
       "<span style=white-space:pre-wrap>        label.ref &lt;- ref.obj@meta.data[, ref.ctype]</span>\n",
       "<span style=white-space:pre-wrap>    else label.ref &lt;- Idents(ref.obj)</span>\n",
       "<span style=white-space:pre-wrap>    ref.df &lt;- cbind.data.frame(ref.df, label = label.ref %&gt;% </span>\n",
       "<span style=white-space:pre-wrap>        as.vector)</span>\n",
       "<span style=white-space:pre-wrap>    res.pred &lt;- scibetR::SciBet_R(ref.df, test.df, ...)</span>\n",
       "<span style=white-space:pre-wrap>    obj.seu$sciBet &lt;- res.pred</span>\n",
       "<span style=white-space:pre-wrap>    return(obj.seu)</span>\n",
       "}</code></pre></dd>\n",
       "\t<dt>$scmap</dt>\n",
       "\t\t<dd><pre class=language-r><code>function (obj.seu, ref.obj, ref.ctype = NULL, slot.data = c(\"data\", \n",
       "<span style=white-space:pre-wrap>    \"count\"), ...) </span>\n",
       "{\n",
       "<span style=white-space:pre-wrap>    system.file(\"modules\", \"scmap-master\", package = \"scAnnoX\") %&gt;% </span>\n",
       "<span style=white-space:pre-wrap>        load_all(.)</span>\n",
       "<span style=white-space:pre-wrap>    ref.expr &lt;- GetAssayData(ref.obj, slot = slot.data) %&gt;% as.matrix</span>\n",
       "<span style=white-space:pre-wrap>    if (!is.null(ref.ctype)) </span>\n",
       "<span style=white-space:pre-wrap>        label.ref &lt;- ref.obj@meta.data[, ref.ctype]</span>\n",
       "<span style=white-space:pre-wrap>    else label.ref &lt;- Idents(ref.obj)</span>\n",
       "<span style=white-space:pre-wrap>    sce &lt;- SingleCellExperiment(assays = list(normcounts = as.matrix(ref.expr)), </span>\n",
       "<span style=white-space:pre-wrap>        colData = label.ref)</span>\n",
       "<span style=white-space:pre-wrap>    if (match.arg(slot.data) == \"count\") </span>\n",
       "<span style=white-space:pre-wrap>        logcounts(sce) &lt;- log2(normcounts(sce) + 1)</span>\n",
       "<span style=white-space:pre-wrap>    else logcounts(sce) &lt;- normcounts(sce)</span>\n",
       "<span style=white-space:pre-wrap>    rowData(sce)$feature_symbol &lt;- rownames(sce)</span>\n",
       "<span style=white-space:pre-wrap>    sce &lt;- sce[!duplicated(rownames(sce)), ] %&gt;% selectFeatures(.) %&gt;% </span>\n",
       "<span style=white-space:pre-wrap>        indexCell(.)</span>\n",
       "<span style=white-space:pre-wrap>    test.expr &lt;- GetAssayData(obj.seu, slot = slot.data) %&gt;% </span>\n",
       "<span style=white-space:pre-wrap>        as.matrix</span>\n",
       "<span style=white-space:pre-wrap>    sce.test &lt;- SingleCellExperiment(assays = list(normcounts = as.matrix(test.expr)))</span>\n",
       "<span style=white-space:pre-wrap>    if (match.arg(slot.data) == \"count\") </span>\n",
       "<span style=white-space:pre-wrap>        logcounts(sce.test) &lt;- log2(normcounts(sce.test) + 1)</span>\n",
       "<span style=white-space:pre-wrap>    else logcounts(sce.test) &lt;- normcounts(sce.test)</span>\n",
       "<span style=white-space:pre-wrap>    rowData(sce.test)$feature_symbol &lt;- rownames(sce.test)</span>\n",
       "<span style=white-space:pre-wrap>    sce.test &lt;- sce.test[!duplicated(rownames(sce.test)), ] %&gt;% </span>\n",
       "<span style=white-space:pre-wrap>        selectFeatures(.) %&gt;% indexCell(.)</span>\n",
       "<span style=white-space:pre-wrap>    pre.res &lt;- scmapCell(sce.test, list(metadata(sce)$scmap_cell_index))</span>\n",
       "<span style=white-space:pre-wrap>    scmap.clusters &lt;- scmapCell2Cluster(pre.res, list(as.character(colData(sce)[, </span>\n",
       "<span style=white-space:pre-wrap>        1])))</span>\n",
       "<span style=white-space:pre-wrap>    obj.seu$scmapAnno &lt;- scmap.clusters$combined_labs</span>\n",
       "<span style=white-space:pre-wrap>    return(obj.seu)</span>\n",
       "}</code></pre></dd>\n",
       "\t<dt>$CHETAH</dt>\n",
       "\t\t<dd><pre class=language-r><code>function (obj.seu, ref.obj, ref.ctype = NULL, slot.data = \"data\", \n",
       "<span style=white-space:pre-wrap>    ...) </span>\n",
       "{\n",
       "<span style=white-space:pre-wrap>    test.expr &lt;- GetAssayData(obj.seu, slot = slot.data) %&gt;% </span>\n",
       "<span style=white-space:pre-wrap>        as.matrix</span>\n",
       "<span style=white-space:pre-wrap>    sce.test &lt;- SingleCellExperiment(assays = list(counts = test.expr))</span>\n",
       "<span style=white-space:pre-wrap>    if (slot.data == \"count\") </span>\n",
       "<span style=white-space:pre-wrap>        sce.test &lt;- scater::logNormCounts(sce.test)</span>\n",
       "<span style=white-space:pre-wrap>    ref.expr &lt;- GetAssayData(ref.obj, slot = slot.data) %&gt;% as.matrix</span>\n",
       "<span style=white-space:pre-wrap>    if (!is.null(ref.ctype)) </span>\n",
       "<span style=white-space:pre-wrap>        label.ref &lt;- ref.obj@meta.data[, ref.ctype]</span>\n",
       "<span style=white-space:pre-wrap>    else label.ref &lt;- Idents(ref.obj)</span>\n",
       "<span style=white-space:pre-wrap>    sce.ref &lt;- SingleCellExperiment(assays = list(counts = as.matrix(ref.expr)), </span>\n",
       "<span style=white-space:pre-wrap>        colData = DataFrame(celltypes = label.ref))</span>\n",
       "<span style=white-space:pre-wrap>    if (slot.data == \"count\") </span>\n",
       "<span style=white-space:pre-wrap>        sce.ref &lt;- scater::logNormCounts(sce.ref)</span>\n",
       "<span style=white-space:pre-wrap>    pred.res &lt;- CHETAH::CHETAHclassifier(input = sce.test, ref_cells = sce.ref, </span>\n",
       "<span style=white-space:pre-wrap>        ...)</span>\n",
       "<span style=white-space:pre-wrap>    obj.seu$CHETAH &lt;- pred.res$celltype_CHETAH %&gt;% data.frame</span>\n",
       "<span style=white-space:pre-wrap>    return(obj.seu)</span>\n",
       "}</code></pre></dd>\n",
       "\t<dt>$scSorter</dt>\n",
       "\t\t<dd><pre class=language-r><code>function (obj.seu, marker.lst, ...) \n",
       "{\n",
       "<span style=white-space:pre-wrap>    test.expr &lt;- GetAssayData(obj.seu, slot = \"data\") %&gt;% as.matrix</span>\n",
       "<span style=white-space:pre-wrap>    anno &lt;- collapseLisToFrame(marker.lst)</span>\n",
       "<span style=white-space:pre-wrap>    pred.res &lt;- scSorter::scSorter(expr = test.expr, anno = anno, </span>\n",
       "<span style=white-space:pre-wrap>        ...)</span>\n",
       "<span style=white-space:pre-wrap>    obj.seu$scSorterAnno &lt;- pred.res$Pred_Type</span>\n",
       "<span style=white-space:pre-wrap>    return(obj.seu)</span>\n",
       "}</code></pre></dd>\n",
       "\t<dt>$scCATCH</dt>\n",
       "\t\t<dd><pre class=language-r><code>function (obj.seu, marker.lst = NULL, qry.cluster = \"idents\", \n",
       "<span style=white-space:pre-wrap>    species = c(\"Human\", \"Mouse\"), ...) </span>\n",
       "{\n",
       "<span style=white-space:pre-wrap>    test.expr &lt;- GetAssayData(obj.seu, slot = \"data\") %&gt;% as.matrix</span>\n",
       "<span style=white-space:pre-wrap>    if (qry.cluster == \"idents\") </span>\n",
       "<span style=white-space:pre-wrap>        clus.info &lt;- Idents(obj.seu)</span>\n",
       "<span style=white-space:pre-wrap>    else clus.info &lt;- obj.seu@meta.data[, qry.cluster]</span>\n",
       "<span style=white-space:pre-wrap>    data &lt;- scCATCH::createscCATCH(data = test.expr, cluster = clus.info %&gt;% </span>\n",
       "<span style=white-space:pre-wrap>        as.vector)</span>\n",
       "<span style=white-space:pre-wrap>    if (is.null(marker.lst)) {</span>\n",
       "<span style=white-space:pre-wrap>        cellmatch &lt;- scCATCH::cellmatch</span>\n",
       "<span style=white-space:pre-wrap>        marker.db &lt;- cellmatch[cellmatch$species == match.arg(species), </span>\n",
       "<span style=white-space:pre-wrap>            ]</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    else {</span>\n",
       "<span style=white-space:pre-wrap>        marker.db &lt;- collapseLisToFrame(marker.lst) %&gt;% `colnames&lt;-`(c(\"celltype\", </span>\n",
       "<span style=white-space:pre-wrap>            \"gene\")) %&gt;% mutate(species = \"Human\", tissue = NA, </span>\n",
       "<span style=white-space:pre-wrap>            cancer = \"Normal\", condition = \"Normal\", subtype1 = NA, </span>\n",
       "<span style=white-space:pre-wrap>            subtype2 = NA, subtype3 = NA, resource = NA, pmid = NA)</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    obj &lt;- scCATCH::findmarkergene(data, species = match.arg(species), </span>\n",
       "<span style=white-space:pre-wrap>        if_use_custom_marker = TRUE, marker = marker.db)</span>\n",
       "<span style=white-space:pre-wrap>    pred.res &lt;- scCATCH::findcelltype(obj)</span>\n",
       "<span style=white-space:pre-wrap>    obj.tmp &lt;- RenameIdents(obj.seu, split(pred.res@celltype$cell_type, </span>\n",
       "<span style=white-space:pre-wrap>        pred.res@celltype$cluster) %&gt;% unlist(.))</span>\n",
       "<span style=white-space:pre-wrap>    obj.seu$scCATCH &lt;- Idents(obj.tmp) %&gt;% as.vector</span>\n",
       "<span style=white-space:pre-wrap>    return(obj.seu)</span>\n",
       "}</code></pre></dd>\n",
       "\t<dt>$cellID</dt>\n",
       "\t\t<dd><pre class=language-r><code>function (obj.seu, marker.lst = NULL, gset.len = 5, sig.cut = 2, \n",
       "<span style=white-space:pre-wrap>    species = c(\"Hs\", \"Mm\"), tissue.type = NULL, ...) </span>\n",
       "{\n",
       "<span style=white-space:pre-wrap>    assay.names &lt;- obj.seu@assays %&gt;% names</span>\n",
       "<span style=white-space:pre-wrap>    if (!(\"data\" %in% assay.names)) </span>\n",
       "<span style=white-space:pre-wrap>        obj.seu@assays$data &lt;- obj.seu@assays$RNA</span>\n",
       "<span style=white-space:pre-wrap>    obj.seu &lt;- CelliD::RunMCA(obj.seu)</span>\n",
       "<span style=white-space:pre-wrap>    if (is.null(marker.lst)) {</span>\n",
       "<span style=white-space:pre-wrap>        panglao &lt;- system.file(\"data\", \"PanglaoDB_markers_27_Mar_2020.tsv.gz\", </span>\n",
       "<span style=white-space:pre-wrap>            package = \"scAnnoX\") %&gt;% read_tsv</span>\n",
       "<span style=white-space:pre-wrap>        panglao.all &lt;- panglao %&gt;% filter(str_detect(species, </span>\n",
       "<span style=white-space:pre-wrap>            match.arg(species)))</span>\n",
       "<span style=white-space:pre-wrap>        if (!is.null(tissue.type)) </span>\n",
       "<span style=white-space:pre-wrap>            panglao.all &lt;- panglao.all %&gt;% filter(organ == tissue.type)</span>\n",
       "<span style=white-space:pre-wrap>        panglao.all &lt;- panglao.all %&gt;% group_by(`cell type`) %&gt;% </span>\n",
       "<span style=white-space:pre-wrap>            summarise(geneset = list(`official gene symbol`))</span>\n",
       "<span style=white-space:pre-wrap>        all.gs &lt;- setNames(panglao.all$geneset, panglao.all$`cell type`)</span>\n",
       "<span style=white-space:pre-wrap>        all.gs &lt;- all.gs[sapply(all.gs, length) &gt;= gset.len]</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    else all.gs &lt;- marker.lst[sapply(marker.lst, length) &gt;= gset.len]</span>\n",
       "<span style=white-space:pre-wrap>    HGT.all.gs &lt;- CelliD::RunCellHGT(obj.seu, pathways = all.gs, </span>\n",
       "<span style=white-space:pre-wrap>        dims = 1:50)</span>\n",
       "<span style=white-space:pre-wrap>    all.gs.pred &lt;- rownames(HGT.all.gs)[apply(HGT.all.gs, 2, </span>\n",
       "<span style=white-space:pre-wrap>        which.max)]</span>\n",
       "<span style=white-space:pre-wrap>    all.gs.pred.signif &lt;- ifelse(apply(HGT.all.gs, 2, max) &gt; </span>\n",
       "<span style=white-space:pre-wrap>        2, yes = all.gs.pred, \"unassigned\")</span>\n",
       "<span style=white-space:pre-wrap>    obj.seu$CellIDAnno &lt;- all.gs.pred</span>\n",
       "<span style=white-space:pre-wrap>    obj.seu$CellIDAnno.adj &lt;- all.gs.pred.signif</span>\n",
       "<span style=white-space:pre-wrap>    return(obj.seu)</span>\n",
       "}</code></pre></dd>\n",
       "\t<dt>$sc.type</dt>\n",
       "\t\t<dd><pre class=language-r><code>function (obj.seu, marker.lst = NULL, tissue.type = \"Immune system\", \n",
       "<span style=white-space:pre-wrap>    ...) </span>\n",
       "{\n",
       "<span style=white-space:pre-wrap>    system.file(\"modules\", \"sc-type-master/R/auto_detect_tissue_type.R\", </span>\n",
       "<span style=white-space:pre-wrap>        package = \"scAnnoX\") %&gt;% source(.)</span>\n",
       "<span style=white-space:pre-wrap>    system.file(\"modules\", \"sc-type-master/R/gene_sets_prepare.R\", </span>\n",
       "<span style=white-space:pre-wrap>        package = \"scAnnoX\") %&gt;% source(.)</span>\n",
       "<span style=white-space:pre-wrap>    system.file(\"modules\", \"sc-type-master/R/sctype_score_.R\", </span>\n",
       "<span style=white-space:pre-wrap>        package = \"scAnnoX\") %&gt;% source(.)</span>\n",
       "<span style=white-space:pre-wrap>    marker.dbfile &lt;- system.file(\"modules\", \"sc-type-master/ScTypeDB_full.xlsx\", </span>\n",
       "<span style=white-space:pre-wrap>        package = \"scAnnoX\")</span>\n",
       "<span style=white-space:pre-wrap>    if (is.null(marker.lst)) {</span>\n",
       "<span style=white-space:pre-wrap>        gs.list &lt;- gene_sets_prepare(marker.dbfile, tissue.type)</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    else {</span>\n",
       "<span style=white-space:pre-wrap>        gs.neg &lt;- lapply(names(marker.lst), function(xx) character(0)) %&gt;% </span>\n",
       "<span style=white-space:pre-wrap>            `names&lt;-`(names(marker.lst))</span>\n",
       "<span style=white-space:pre-wrap>        gs.pos &lt;- marker.lst</span>\n",
       "<span style=white-space:pre-wrap>        gs.list &lt;- list(gs_positive = gs.pos, gs_negative = gs.neg)</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    sc.expr &lt;- GetAssayData(obj.seu) %&gt;% as.matrix</span>\n",
       "<span style=white-space:pre-wrap>    es.max &lt;- sctype_score(scRNAseqData = sc.expr, gs = gs.list$gs_positive, </span>\n",
       "<span style=white-space:pre-wrap>        gs2 = gs.list$gs_negative, ...)</span>\n",
       "<span style=white-space:pre-wrap>    pred.res &lt;- apply(es.max, 2, which.max) %&gt;% rownames(es.max)[.]</span>\n",
       "<span style=white-space:pre-wrap>    obj.seu$scTypeAnno &lt;- pred.res</span>\n",
       "<span style=white-space:pre-wrap>    return(obj.seu)</span>\n",
       "}</code></pre></dd>\n",
       "\t<dt>$SCINA</dt>\n",
       "\t\t<dd><pre class=language-r><code>function (obj.seu, marker.lst, ...) \n",
       "{\n",
       "<span style=white-space:pre-wrap>    pred.res &lt;- SCINA::SCINA(obj.seu %&gt;% GetAssayData(., slot = \"data\") %&gt;% </span>\n",
       "<span style=white-space:pre-wrap>        as.matrix, marker.lst, max_iter = 1000, convergence_n = 12, </span>\n",
       "<span style=white-space:pre-wrap>        convergence_rate = 0.999, sensitivity_cutoff = 0.9, ...)</span>\n",
       "<span style=white-space:pre-wrap>    obj.seu$SCINA &lt;- pred.res$cell_labels</span>\n",
       "<span style=white-space:pre-wrap>    return(obj.seu)</span>\n",
       "}</code></pre></dd>\n",
       "</dl>\n"
      ],
      "text/latex": [
       "\\begin{description}\n",
       "\\item[\\$SingleR] \\begin{minted}{r}\n",
       "function (obj.seu, ref.obj, ref.ctype = NULL, ...) \n",
       "\\{\n",
       "    test.expr <- GetAssayData(obj.seu, slot = \"data\") \\%>\\% as.matrix\n",
       "    ref.expr <- GetAssayData(ref.obj, slot = \"data\") \\%>\\% as.matrix\n",
       "    if (!is.null(ref.ctype)) \n",
       "        label.ref <- ref.obj@meta.data{[}, ref.ctype{]}\n",
       "    else label.ref <- Idents(ref.obj)\n",
       "    res.pred <- SingleR::SingleR(test = test.expr, ref = ref.expr, \n",
       "        labels = label.ref, ...)\n",
       "    obj.seu\\$SingleR <- res.pred\\$labels\n",
       "    return(obj.seu)\n",
       "\\}\n",
       "\\end{minted}\n",
       "\\item[\\$Seurat] \\begin{minted}{r}\n",
       "function (obj.seu, ref.obj, ref.ctype = NULL, npcs = 30, ...) \n",
       "\\{\n",
       "    anchors <- FindTransferAnchors(reference = ref.obj, query = obj.seu, \n",
       "        dims = 1:npcs)\n",
       "    if (!is.null(ref.ctype)) \n",
       "        label.ref <- ref.obj@meta.data{[}, ref.ctype{]}\n",
       "    else label.ref <- Idents(ref.obj)\n",
       "    res.pred <- TransferData(anchorset = anchors, refdata = label.ref, \n",
       "        dims = 1:npcs, ...)\n",
       "    obj.seu\\$SeuratAnno <- res.pred{[}, 1{]}\n",
       "    return(obj.seu)\n",
       "\\}\n",
       "\\end{minted}\n",
       "\\item[\\$sciBet] \\begin{minted}{r}\n",
       "function (obj.seu, ref.obj, ref.ctype = NULL, ...) \n",
       "\\{\n",
       "    test.df <- GetAssayData(obj.seu, slot = \"data\") \\%>\\% as.data.frame \\%>\\% \n",
       "        t\n",
       "    ref.df <- GetAssayData(ref.obj, slot = \"data\") \\%>\\% as.data.frame \\%>\\% \n",
       "        t\n",
       "    if (!is.null(ref.ctype)) \n",
       "        label.ref <- ref.obj@meta.data{[}, ref.ctype{]}\n",
       "    else label.ref <- Idents(ref.obj)\n",
       "    ref.df <- cbind.data.frame(ref.df, label = label.ref \\%>\\% \n",
       "        as.vector)\n",
       "    res.pred <- scibetR::SciBet\\_R(ref.df, test.df, ...)\n",
       "    obj.seu\\$sciBet <- res.pred\n",
       "    return(obj.seu)\n",
       "\\}\n",
       "\\end{minted}\n",
       "\\item[\\$scmap] \\begin{minted}{r}\n",
       "function (obj.seu, ref.obj, ref.ctype = NULL, slot.data = c(\"data\", \n",
       "    \"count\"), ...) \n",
       "\\{\n",
       "    system.file(\"modules\", \"scmap-master\", package = \"scAnnoX\") \\%>\\% \n",
       "        load\\_all(.)\n",
       "    ref.expr <- GetAssayData(ref.obj, slot = slot.data) \\%>\\% as.matrix\n",
       "    if (!is.null(ref.ctype)) \n",
       "        label.ref <- ref.obj@meta.data{[}, ref.ctype{]}\n",
       "    else label.ref <- Idents(ref.obj)\n",
       "    sce <- SingleCellExperiment(assays = list(normcounts = as.matrix(ref.expr)), \n",
       "        colData = label.ref)\n",
       "    if (match.arg(slot.data) == \"count\") \n",
       "        logcounts(sce) <- log2(normcounts(sce) + 1)\n",
       "    else logcounts(sce) <- normcounts(sce)\n",
       "    rowData(sce)\\$feature\\_symbol <- rownames(sce)\n",
       "    sce <- sce{[}!duplicated(rownames(sce)), {]} \\%>\\% selectFeatures(.) \\%>\\% \n",
       "        indexCell(.)\n",
       "    test.expr <- GetAssayData(obj.seu, slot = slot.data) \\%>\\% \n",
       "        as.matrix\n",
       "    sce.test <- SingleCellExperiment(assays = list(normcounts = as.matrix(test.expr)))\n",
       "    if (match.arg(slot.data) == \"count\") \n",
       "        logcounts(sce.test) <- log2(normcounts(sce.test) + 1)\n",
       "    else logcounts(sce.test) <- normcounts(sce.test)\n",
       "    rowData(sce.test)\\$feature\\_symbol <- rownames(sce.test)\n",
       "    sce.test <- sce.test{[}!duplicated(rownames(sce.test)), {]} \\%>\\% \n",
       "        selectFeatures(.) \\%>\\% indexCell(.)\n",
       "    pre.res <- scmapCell(sce.test, list(metadata(sce)\\$scmap\\_cell\\_index))\n",
       "    scmap.clusters <- scmapCell2Cluster(pre.res, list(as.character(colData(sce){[}, \n",
       "        1{]})))\n",
       "    obj.seu\\$scmapAnno <- scmap.clusters\\$combined\\_labs\n",
       "    return(obj.seu)\n",
       "\\}\n",
       "\\end{minted}\n",
       "\\item[\\$CHETAH] \\begin{minted}{r}\n",
       "function (obj.seu, ref.obj, ref.ctype = NULL, slot.data = \"data\", \n",
       "    ...) \n",
       "\\{\n",
       "    test.expr <- GetAssayData(obj.seu, slot = slot.data) \\%>\\% \n",
       "        as.matrix\n",
       "    sce.test <- SingleCellExperiment(assays = list(counts = test.expr))\n",
       "    if (slot.data == \"count\") \n",
       "        sce.test <- scater::logNormCounts(sce.test)\n",
       "    ref.expr <- GetAssayData(ref.obj, slot = slot.data) \\%>\\% as.matrix\n",
       "    if (!is.null(ref.ctype)) \n",
       "        label.ref <- ref.obj@meta.data{[}, ref.ctype{]}\n",
       "    else label.ref <- Idents(ref.obj)\n",
       "    sce.ref <- SingleCellExperiment(assays = list(counts = as.matrix(ref.expr)), \n",
       "        colData = DataFrame(celltypes = label.ref))\n",
       "    if (slot.data == \"count\") \n",
       "        sce.ref <- scater::logNormCounts(sce.ref)\n",
       "    pred.res <- CHETAH::CHETAHclassifier(input = sce.test, ref\\_cells = sce.ref, \n",
       "        ...)\n",
       "    obj.seu\\$CHETAH <- pred.res\\$celltype\\_CHETAH \\%>\\% data.frame\n",
       "    return(obj.seu)\n",
       "\\}\n",
       "\\end{minted}\n",
       "\\item[\\$scSorter] \\begin{minted}{r}\n",
       "function (obj.seu, marker.lst, ...) \n",
       "\\{\n",
       "    test.expr <- GetAssayData(obj.seu, slot = \"data\") \\%>\\% as.matrix\n",
       "    anno <- collapseLisToFrame(marker.lst)\n",
       "    pred.res <- scSorter::scSorter(expr = test.expr, anno = anno, \n",
       "        ...)\n",
       "    obj.seu\\$scSorterAnno <- pred.res\\$Pred\\_Type\n",
       "    return(obj.seu)\n",
       "\\}\n",
       "\\end{minted}\n",
       "\\item[\\$scCATCH] \\begin{minted}{r}\n",
       "function (obj.seu, marker.lst = NULL, qry.cluster = \"idents\", \n",
       "    species = c(\"Human\", \"Mouse\"), ...) \n",
       "\\{\n",
       "    test.expr <- GetAssayData(obj.seu, slot = \"data\") \\%>\\% as.matrix\n",
       "    if (qry.cluster == \"idents\") \n",
       "        clus.info <- Idents(obj.seu)\n",
       "    else clus.info <- obj.seu@meta.data{[}, qry.cluster{]}\n",
       "    data <- scCATCH::createscCATCH(data = test.expr, cluster = clus.info \\%>\\% \n",
       "        as.vector)\n",
       "    if (is.null(marker.lst)) \\{\n",
       "        cellmatch <- scCATCH::cellmatch\n",
       "        marker.db <- cellmatch{[}cellmatch\\$species == match.arg(species), \n",
       "            {]}\n",
       "    \\}\n",
       "    else \\{\n",
       "        marker.db <- collapseLisToFrame(marker.lst) \\%>\\% `colnames<-`(c(\"celltype\", \n",
       "            \"gene\")) \\%>\\% mutate(species = \"Human\", tissue = NA, \n",
       "            cancer = \"Normal\", condition = \"Normal\", subtype1 = NA, \n",
       "            subtype2 = NA, subtype3 = NA, resource = NA, pmid = NA)\n",
       "    \\}\n",
       "    obj <- scCATCH::findmarkergene(data, species = match.arg(species), \n",
       "        if\\_use\\_custom\\_marker = TRUE, marker = marker.db)\n",
       "    pred.res <- scCATCH::findcelltype(obj)\n",
       "    obj.tmp <- RenameIdents(obj.seu, split(pred.res@celltype\\$cell\\_type, \n",
       "        pred.res@celltype\\$cluster) \\%>\\% unlist(.))\n",
       "    obj.seu\\$scCATCH <- Idents(obj.tmp) \\%>\\% as.vector\n",
       "    return(obj.seu)\n",
       "\\}\n",
       "\\end{minted}\n",
       "\\item[\\$cellID] \\begin{minted}{r}\n",
       "function (obj.seu, marker.lst = NULL, gset.len = 5, sig.cut = 2, \n",
       "    species = c(\"Hs\", \"Mm\"), tissue.type = NULL, ...) \n",
       "\\{\n",
       "    assay.names <- obj.seu@assays \\%>\\% names\n",
       "    if (!(\"data\" \\%in\\% assay.names)) \n",
       "        obj.seu@assays\\$data <- obj.seu@assays\\$RNA\n",
       "    obj.seu <- CelliD::RunMCA(obj.seu)\n",
       "    if (is.null(marker.lst)) \\{\n",
       "        panglao <- system.file(\"data\", \"PanglaoDB\\_markers\\_27\\_Mar\\_2020.tsv.gz\", \n",
       "            package = \"scAnnoX\") \\%>\\% read\\_tsv\n",
       "        panglao.all <- panglao \\%>\\% filter(str\\_detect(species, \n",
       "            match.arg(species)))\n",
       "        if (!is.null(tissue.type)) \n",
       "            panglao.all <- panglao.all \\%>\\% filter(organ == tissue.type)\n",
       "        panglao.all <- panglao.all \\%>\\% group\\_by(`cell type`) \\%>\\% \n",
       "            summarise(geneset = list(`official gene symbol`))\n",
       "        all.gs <- setNames(panglao.all\\$geneset, panglao.all\\$`cell type`)\n",
       "        all.gs <- all.gs{[}sapply(all.gs, length) >= gset.len{]}\n",
       "    \\}\n",
       "    else all.gs <- marker.lst{[}sapply(marker.lst, length) >= gset.len{]}\n",
       "    HGT.all.gs <- CelliD::RunCellHGT(obj.seu, pathways = all.gs, \n",
       "        dims = 1:50)\n",
       "    all.gs.pred <- rownames(HGT.all.gs){[}apply(HGT.all.gs, 2, \n",
       "        which.max){]}\n",
       "    all.gs.pred.signif <- ifelse(apply(HGT.all.gs, 2, max) > \n",
       "        2, yes = all.gs.pred, \"unassigned\")\n",
       "    obj.seu\\$CellIDAnno <- all.gs.pred\n",
       "    obj.seu\\$CellIDAnno.adj <- all.gs.pred.signif\n",
       "    return(obj.seu)\n",
       "\\}\n",
       "\\end{minted}\n",
       "\\item[\\$sc.type] \\begin{minted}{r}\n",
       "function (obj.seu, marker.lst = NULL, tissue.type = \"Immune system\", \n",
       "    ...) \n",
       "\\{\n",
       "    system.file(\"modules\", \"sc-type-master/R/auto\\_detect\\_tissue\\_type.R\", \n",
       "        package = \"scAnnoX\") \\%>\\% source(.)\n",
       "    system.file(\"modules\", \"sc-type-master/R/gene\\_sets\\_prepare.R\", \n",
       "        package = \"scAnnoX\") \\%>\\% source(.)\n",
       "    system.file(\"modules\", \"sc-type-master/R/sctype\\_score\\_.R\", \n",
       "        package = \"scAnnoX\") \\%>\\% source(.)\n",
       "    marker.dbfile <- system.file(\"modules\", \"sc-type-master/ScTypeDB\\_full.xlsx\", \n",
       "        package = \"scAnnoX\")\n",
       "    if (is.null(marker.lst)) \\{\n",
       "        gs.list <- gene\\_sets\\_prepare(marker.dbfile, tissue.type)\n",
       "    \\}\n",
       "    else \\{\n",
       "        gs.neg <- lapply(names(marker.lst), function(xx) character(0)) \\%>\\% \n",
       "            `names<-`(names(marker.lst))\n",
       "        gs.pos <- marker.lst\n",
       "        gs.list <- list(gs\\_positive = gs.pos, gs\\_negative = gs.neg)\n",
       "    \\}\n",
       "    sc.expr <- GetAssayData(obj.seu) \\%>\\% as.matrix\n",
       "    es.max <- sctype\\_score(scRNAseqData = sc.expr, gs = gs.list\\$gs\\_positive, \n",
       "        gs2 = gs.list\\$gs\\_negative, ...)\n",
       "    pred.res <- apply(es.max, 2, which.max) \\%>\\% rownames(es.max){[}.{]}\n",
       "    obj.seu\\$scTypeAnno <- pred.res\n",
       "    return(obj.seu)\n",
       "\\}\n",
       "\\end{minted}\n",
       "\\item[\\$SCINA] \\begin{minted}{r}\n",
       "function (obj.seu, marker.lst, ...) \n",
       "\\{\n",
       "    pred.res <- SCINA::SCINA(obj.seu \\%>\\% GetAssayData(., slot = \"data\") \\%>\\% \n",
       "        as.matrix, marker.lst, max\\_iter = 1000, convergence\\_n = 12, \n",
       "        convergence\\_rate = 0.999, sensitivity\\_cutoff = 0.9, ...)\n",
       "    obj.seu\\$SCINA <- pred.res\\$cell\\_labels\n",
       "    return(obj.seu)\n",
       "\\}\n",
       "\\end{minted}\n",
       "\\end{description}\n"
      ],
      "text/markdown": [
       "$SingleR\n",
       ":   ```r\n",
       "function (obj.seu, ref.obj, ref.ctype = NULL, ...) \n",
       "{\n",
       "    test.expr <- GetAssayData(obj.seu, slot = \"data\") %>% as.matrix\n",
       "    ref.expr <- GetAssayData(ref.obj, slot = \"data\") %>% as.matrix\n",
       "    if (!is.null(ref.ctype)) \n",
       "        label.ref <- ref.obj@meta.data[, ref.ctype]\n",
       "    else label.ref <- Idents(ref.obj)\n",
       "    res.pred <- SingleR::SingleR(test = test.expr, ref = ref.expr, \n",
       "        labels = label.ref, ...)\n",
       "    obj.seu$SingleR <- res.pred$labels\n",
       "    return(obj.seu)\n",
       "}\n",
       "```\n",
       "$Seurat\n",
       ":   ```r\n",
       "function (obj.seu, ref.obj, ref.ctype = NULL, npcs = 30, ...) \n",
       "{\n",
       "    anchors <- FindTransferAnchors(reference = ref.obj, query = obj.seu, \n",
       "        dims = 1:npcs)\n",
       "    if (!is.null(ref.ctype)) \n",
       "        label.ref <- ref.obj@meta.data[, ref.ctype]\n",
       "    else label.ref <- Idents(ref.obj)\n",
       "    res.pred <- TransferData(anchorset = anchors, refdata = label.ref, \n",
       "        dims = 1:npcs, ...)\n",
       "    obj.seu$SeuratAnno <- res.pred[, 1]\n",
       "    return(obj.seu)\n",
       "}\n",
       "```\n",
       "$sciBet\n",
       ":   ```r\n",
       "function (obj.seu, ref.obj, ref.ctype = NULL, ...) \n",
       "{\n",
       "    test.df <- GetAssayData(obj.seu, slot = \"data\") %>% as.data.frame %>% \n",
       "        t\n",
       "    ref.df <- GetAssayData(ref.obj, slot = \"data\") %>% as.data.frame %>% \n",
       "        t\n",
       "    if (!is.null(ref.ctype)) \n",
       "        label.ref <- ref.obj@meta.data[, ref.ctype]\n",
       "    else label.ref <- Idents(ref.obj)\n",
       "    ref.df <- cbind.data.frame(ref.df, label = label.ref %>% \n",
       "        as.vector)\n",
       "    res.pred <- scibetR::SciBet_R(ref.df, test.df, ...)\n",
       "    obj.seu$sciBet <- res.pred\n",
       "    return(obj.seu)\n",
       "}\n",
       "```\n",
       "$scmap\n",
       ":   ```r\n",
       "function (obj.seu, ref.obj, ref.ctype = NULL, slot.data = c(\"data\", \n",
       "    \"count\"), ...) \n",
       "{\n",
       "    system.file(\"modules\", \"scmap-master\", package = \"scAnnoX\") %>% \n",
       "        load_all(.)\n",
       "    ref.expr <- GetAssayData(ref.obj, slot = slot.data) %>% as.matrix\n",
       "    if (!is.null(ref.ctype)) \n",
       "        label.ref <- ref.obj@meta.data[, ref.ctype]\n",
       "    else label.ref <- Idents(ref.obj)\n",
       "    sce <- SingleCellExperiment(assays = list(normcounts = as.matrix(ref.expr)), \n",
       "        colData = label.ref)\n",
       "    if (match.arg(slot.data) == \"count\") \n",
       "        logcounts(sce) <- log2(normcounts(sce) + 1)\n",
       "    else logcounts(sce) <- normcounts(sce)\n",
       "    rowData(sce)$feature_symbol <- rownames(sce)\n",
       "    sce <- sce[!duplicated(rownames(sce)), ] %>% selectFeatures(.) %>% \n",
       "        indexCell(.)\n",
       "    test.expr <- GetAssayData(obj.seu, slot = slot.data) %>% \n",
       "        as.matrix\n",
       "    sce.test <- SingleCellExperiment(assays = list(normcounts = as.matrix(test.expr)))\n",
       "    if (match.arg(slot.data) == \"count\") \n",
       "        logcounts(sce.test) <- log2(normcounts(sce.test) + 1)\n",
       "    else logcounts(sce.test) <- normcounts(sce.test)\n",
       "    rowData(sce.test)$feature_symbol <- rownames(sce.test)\n",
       "    sce.test <- sce.test[!duplicated(rownames(sce.test)), ] %>% \n",
       "        selectFeatures(.) %>% indexCell(.)\n",
       "    pre.res <- scmapCell(sce.test, list(metadata(sce)$scmap_cell_index))\n",
       "    scmap.clusters <- scmapCell2Cluster(pre.res, list(as.character(colData(sce)[, \n",
       "        1])))\n",
       "    obj.seu$scmapAnno <- scmap.clusters$combined_labs\n",
       "    return(obj.seu)\n",
       "}\n",
       "```\n",
       "$CHETAH\n",
       ":   ```r\n",
       "function (obj.seu, ref.obj, ref.ctype = NULL, slot.data = \"data\", \n",
       "    ...) \n",
       "{\n",
       "    test.expr <- GetAssayData(obj.seu, slot = slot.data) %>% \n",
       "        as.matrix\n",
       "    sce.test <- SingleCellExperiment(assays = list(counts = test.expr))\n",
       "    if (slot.data == \"count\") \n",
       "        sce.test <- scater::logNormCounts(sce.test)\n",
       "    ref.expr <- GetAssayData(ref.obj, slot = slot.data) %>% as.matrix\n",
       "    if (!is.null(ref.ctype)) \n",
       "        label.ref <- ref.obj@meta.data[, ref.ctype]\n",
       "    else label.ref <- Idents(ref.obj)\n",
       "    sce.ref <- SingleCellExperiment(assays = list(counts = as.matrix(ref.expr)), \n",
       "        colData = DataFrame(celltypes = label.ref))\n",
       "    if (slot.data == \"count\") \n",
       "        sce.ref <- scater::logNormCounts(sce.ref)\n",
       "    pred.res <- CHETAH::CHETAHclassifier(input = sce.test, ref_cells = sce.ref, \n",
       "        ...)\n",
       "    obj.seu$CHETAH <- pred.res$celltype_CHETAH %>% data.frame\n",
       "    return(obj.seu)\n",
       "}\n",
       "```\n",
       "$scSorter\n",
       ":   ```r\n",
       "function (obj.seu, marker.lst, ...) \n",
       "{\n",
       "    test.expr <- GetAssayData(obj.seu, slot = \"data\") %>% as.matrix\n",
       "    anno <- collapseLisToFrame(marker.lst)\n",
       "    pred.res <- scSorter::scSorter(expr = test.expr, anno = anno, \n",
       "        ...)\n",
       "    obj.seu$scSorterAnno <- pred.res$Pred_Type\n",
       "    return(obj.seu)\n",
       "}\n",
       "```\n",
       "$scCATCH\n",
       ":   ```r\n",
       "function (obj.seu, marker.lst = NULL, qry.cluster = \"idents\", \n",
       "    species = c(\"Human\", \"Mouse\"), ...) \n",
       "{\n",
       "    test.expr <- GetAssayData(obj.seu, slot = \"data\") %>% as.matrix\n",
       "    if (qry.cluster == \"idents\") \n",
       "        clus.info <- Idents(obj.seu)\n",
       "    else clus.info <- obj.seu@meta.data[, qry.cluster]\n",
       "    data <- scCATCH::createscCATCH(data = test.expr, cluster = clus.info %>% \n",
       "        as.vector)\n",
       "    if (is.null(marker.lst)) {\n",
       "        cellmatch <- scCATCH::cellmatch\n",
       "        marker.db <- cellmatch[cellmatch$species == match.arg(species), \n",
       "            ]\n",
       "    }\n",
       "    else {\n",
       "        marker.db <- collapseLisToFrame(marker.lst) %>% `colnames<-`(c(\"celltype\", \n",
       "            \"gene\")) %>% mutate(species = \"Human\", tissue = NA, \n",
       "            cancer = \"Normal\", condition = \"Normal\", subtype1 = NA, \n",
       "            subtype2 = NA, subtype3 = NA, resource = NA, pmid = NA)\n",
       "    }\n",
       "    obj <- scCATCH::findmarkergene(data, species = match.arg(species), \n",
       "        if_use_custom_marker = TRUE, marker = marker.db)\n",
       "    pred.res <- scCATCH::findcelltype(obj)\n",
       "    obj.tmp <- RenameIdents(obj.seu, split(pred.res@celltype$cell_type, \n",
       "        pred.res@celltype$cluster) %>% unlist(.))\n",
       "    obj.seu$scCATCH <- Idents(obj.tmp) %>% as.vector\n",
       "    return(obj.seu)\n",
       "}\n",
       "```\n",
       "$cellID\n",
       ":   ```r\n",
       "function (obj.seu, marker.lst = NULL, gset.len = 5, sig.cut = 2, \n",
       "    species = c(\"Hs\", \"Mm\"), tissue.type = NULL, ...) \n",
       "{\n",
       "    assay.names <- obj.seu@assays %>% names\n",
       "    if (!(\"data\" %in% assay.names)) \n",
       "        obj.seu@assays$data <- obj.seu@assays$RNA\n",
       "    obj.seu <- CelliD::RunMCA(obj.seu)\n",
       "    if (is.null(marker.lst)) {\n",
       "        panglao <- system.file(\"data\", \"PanglaoDB_markers_27_Mar_2020.tsv.gz\", \n",
       "            package = \"scAnnoX\") %>% read_tsv\n",
       "        panglao.all <- panglao %>% filter(str_detect(species, \n",
       "            match.arg(species)))\n",
       "        if (!is.null(tissue.type)) \n",
       "            panglao.all <- panglao.all %>% filter(organ == tissue.type)\n",
       "        panglao.all <- panglao.all %>% group_by(`cell type`) %>% \n",
       "            summarise(geneset = list(`official gene symbol`))\n",
       "        all.gs <- setNames(panglao.all$geneset, panglao.all$`cell type`)\n",
       "        all.gs <- all.gs[sapply(all.gs, length) >= gset.len]\n",
       "    }\n",
       "    else all.gs <- marker.lst[sapply(marker.lst, length) >= gset.len]\n",
       "    HGT.all.gs <- CelliD::RunCellHGT(obj.seu, pathways = all.gs, \n",
       "        dims = 1:50)\n",
       "    all.gs.pred <- rownames(HGT.all.gs)[apply(HGT.all.gs, 2, \n",
       "        which.max)]\n",
       "    all.gs.pred.signif <- ifelse(apply(HGT.all.gs, 2, max) > \n",
       "        2, yes = all.gs.pred, \"unassigned\")\n",
       "    obj.seu$CellIDAnno <- all.gs.pred\n",
       "    obj.seu$CellIDAnno.adj <- all.gs.pred.signif\n",
       "    return(obj.seu)\n",
       "}\n",
       "```\n",
       "$sc.type\n",
       ":   ```r\n",
       "function (obj.seu, marker.lst = NULL, tissue.type = \"Immune system\", \n",
       "    ...) \n",
       "{\n",
       "    system.file(\"modules\", \"sc-type-master/R/auto_detect_tissue_type.R\", \n",
       "        package = \"scAnnoX\") %>% source(.)\n",
       "    system.file(\"modules\", \"sc-type-master/R/gene_sets_prepare.R\", \n",
       "        package = \"scAnnoX\") %>% source(.)\n",
       "    system.file(\"modules\", \"sc-type-master/R/sctype_score_.R\", \n",
       "        package = \"scAnnoX\") %>% source(.)\n",
       "    marker.dbfile <- system.file(\"modules\", \"sc-type-master/ScTypeDB_full.xlsx\", \n",
       "        package = \"scAnnoX\")\n",
       "    if (is.null(marker.lst)) {\n",
       "        gs.list <- gene_sets_prepare(marker.dbfile, tissue.type)\n",
       "    }\n",
       "    else {\n",
       "        gs.neg <- lapply(names(marker.lst), function(xx) character(0)) %>% \n",
       "            `names<-`(names(marker.lst))\n",
       "        gs.pos <- marker.lst\n",
       "        gs.list <- list(gs_positive = gs.pos, gs_negative = gs.neg)\n",
       "    }\n",
       "    sc.expr <- GetAssayData(obj.seu) %>% as.matrix\n",
       "    es.max <- sctype_score(scRNAseqData = sc.expr, gs = gs.list$gs_positive, \n",
       "        gs2 = gs.list$gs_negative, ...)\n",
       "    pred.res <- apply(es.max, 2, which.max) %>% rownames(es.max)[.]\n",
       "    obj.seu$scTypeAnno <- pred.res\n",
       "    return(obj.seu)\n",
       "}\n",
       "```\n",
       "$SCINA\n",
       ":   ```r\n",
       "function (obj.seu, marker.lst, ...) \n",
       "{\n",
       "    pred.res <- SCINA::SCINA(obj.seu %>% GetAssayData(., slot = \"data\") %>% \n",
       "        as.matrix, marker.lst, max_iter = 1000, convergence_n = 12, \n",
       "        convergence_rate = 0.999, sensitivity_cutoff = 0.9, ...)\n",
       "    obj.seu$SCINA <- pred.res$cell_labels\n",
       "    return(obj.seu)\n",
       "}\n",
       "```\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "$SingleR\n",
       "function(obj.seu, ref.obj, ref.ctype = NULL, ...) {\n",
       "    test.expr <- GetAssayData(obj.seu, slot = \"data\") %>% as.matrix\n",
       "\tref.expr <- GetAssayData(ref.obj, slot = \"data\") %>% as.matrix\n",
       "\t\n",
       "\tif (!is.null(ref.ctype)) label.ref <- ref.obj@meta.data[, ref.ctype]\n",
       "\telse label.ref <- Idents(ref.obj)\n",
       "    res.pred <- SingleR::SingleR(test = test.expr, ref = ref.expr, labels = label.ref, ...)\n",
       "    obj.seu$SingleR <- res.pred$labels\n",
       "    return(obj.seu)\n",
       "}\n",
       "<environment: namespace:scAnnoX>\n",
       "\n",
       "$Seurat\n",
       "function(obj.seu, ref.obj, ref.ctype = NULL, npcs = 30, ...) { \n",
       "\t anchors <- FindTransferAnchors(reference = ref.obj, query = obj.seu, dims = 1 : npcs)\n",
       "     if (!is.null(ref.ctype)) label.ref <- ref.obj@meta.data[, ref.ctype]\n",
       "     else label.ref <- Idents(ref.obj)\n",
       "     res.pred <- TransferData(anchorset = anchors, refdata = label.ref,dims = 1 : npcs, ...)\n",
       "     obj.seu$SeuratAnno <- res.pred[, 1]\n",
       "     return(obj.seu)\n",
       "}\n",
       "<environment: namespace:scAnnoX>\n",
       "\n",
       "$sciBet\n",
       "function(obj.seu, ref.obj, ref.ctype = NULL, ...) {\n",
       "\ttest.df <- GetAssayData(obj.seu, slot = \"data\") %>% as.data.frame %>% t\n",
       "    \tref.df <- GetAssayData(ref.obj, slot = \"data\") %>% as.data.frame %>% t\n",
       "    \tif (!is.null(ref.ctype)) label.ref <- ref.obj@meta.data[, ref.ctype]\n",
       "    \telse label.ref <- Idents(ref.obj)\n",
       "\tref.df <- cbind.data.frame(ref.df, label = label.ref %>% as.vector)\n",
       "\tres.pred <- scibetR::SciBet_R(ref.df, test.df, ...)\n",
       "\tobj.seu$sciBet <- res.pred\n",
       "\treturn(obj.seu)\n",
       "}\n",
       "<environment: namespace:scAnnoX>\n",
       "\n",
       "$scmap\n",
       "function(obj.seu, ref.obj, ref.ctype = NULL, slot.data = c('data', 'count'), ...) {\n",
       "    system.file('modules', 'scmap-master', package = 'scAnnoX') %>% load_all(.)\n",
       "\t\n",
       "    ref.expr <- GetAssayData(ref.obj, slot = slot.data) %>% as.matrix\n",
       "    if (!is.null(ref.ctype)) label.ref <- ref.obj@meta.data[, ref.ctype]\n",
       "\telse label.ref <- Idents(ref.obj)\n",
       "\tsce <- SingleCellExperiment(assays = list(normcounts = as.matrix(ref.expr)), colData = label.ref)\n",
       "    \n",
       "\tif (match.arg(slot.data) == 'count') logcounts(sce) <- log2(normcounts(sce) + 1)\n",
       "\telse logcounts(sce) <- normcounts(sce)\n",
       "\trowData(sce)$feature_symbol <- rownames(sce)\n",
       "    sce <- sce[!duplicated(rownames(sce)), ] %>% selectFeatures(.) %>% indexCell(.)\n",
       " \n",
       "    test.expr <- GetAssayData(obj.seu, slot = slot.data) %>% as.matrix\n",
       "    sce.test <- SingleCellExperiment(assays = list(normcounts = as.matrix(test.expr)))\n",
       "    if (match.arg(slot.data) == 'count') logcounts(sce.test) <- log2(normcounts(sce.test) + 1)\n",
       "\telse logcounts(sce.test) <- normcounts(sce.test)\n",
       "    rowData(sce.test)$feature_symbol <- rownames(sce.test)\n",
       "    sce.test <- sce.test[!duplicated(rownames(sce.test)), ] %>% selectFeatures(.) %>% indexCell(.)\n",
       "\n",
       "    pre.res <- scmapCell(sce.test, list(metadata(sce)$scmap_cell_index))\n",
       "    scmap.clusters <- scmapCell2Cluster(pre.res, list(as.character(colData(sce)[, 1])))\n",
       "    obj.seu$scmapAnno <- scmap.clusters$combined_labs\n",
       "    return(obj.seu)\n",
       "}\n",
       "<environment: namespace:scAnnoX>\n",
       "\n",
       "$CHETAH\n",
       "function(obj.seu, ref.obj, ref.ctype = NULL, slot.data = 'data', ...) {\n",
       "\ttest.expr <- GetAssayData(obj.seu, slot = slot.data) %>% as.matrix\n",
       "\tsce.test <- SingleCellExperiment(assays = list(counts = test.expr))\n",
       "\tif (slot.data == 'count') sce.test <- scater::logNormCounts(sce.test)\n",
       "\n",
       "\tref.expr <- GetAssayData(ref.obj, slot = slot.data) %>% as.matrix\n",
       "\tif (!is.null(ref.ctype)) label.ref <- ref.obj@meta.data[, ref.ctype]\n",
       "\telse label.ref <- Idents(ref.obj)\n",
       "\tsce.ref <- SingleCellExperiment(\n",
       "\t\tassays = list(counts = as.matrix(ref.expr)),\n",
       "\t\tcolData = DataFrame(celltypes = label.ref)\n",
       "\t)\n",
       "\tif (slot.data == 'count') sce.ref <- scater::logNormCounts(sce.ref)\n",
       "\tpred.res <- CHETAH::CHETAHclassifier(input = sce.test, ref_cells = sce.ref, ...)\n",
       "\tobj.seu$CHETAH <- pred.res$celltype_CHETAH %>% data.frame\n",
       "\treturn(obj.seu)\n",
       "}\n",
       "<environment: namespace:scAnnoX>\n",
       "\n",
       "$scSorter\n",
       "function(obj.seu, marker.lst, ...) {\n",
       "    test.expr <- GetAssayData(obj.seu, slot = \"data\") %>% as.matrix\n",
       "    anno <- collapseLisToFrame(marker.lst)\n",
       "\tpred.res <- scSorter::scSorter(expr = test.expr, anno = anno, ...)\n",
       "\tobj.seu$scSorterAnno <- pred.res$Pred_Type\n",
       "\treturn(obj.seu)\n",
       "}\n",
       "<environment: namespace:scAnnoX>\n",
       "\n",
       "$scCATCH\n",
       "function(obj.seu, marker.lst = NULL, qry.cluster = 'idents', species = c('Human', 'Mouse'), ...) {\n",
       "    test.expr <- GetAssayData(obj.seu, slot = \"data\") %>% as.matrix\n",
       "\tif (qry.cluster == 'idents') clus.info <- Idents(obj.seu)\n",
       "\telse clus.info <- obj.seu@meta.data[, qry.cluster]\n",
       "\tdata <- scCATCH::createscCATCH(data = test.expr, cluster = clus.info %>% as.vector)\n",
       "    \n",
       "\tif (is.null(marker.lst)) {\n",
       "\t\tcellmatch <- scCATCH::cellmatch\n",
       "\t\tmarker.db <- cellmatch[cellmatch$species == match.arg(species), ]\n",
       "\t} else {\n",
       "\t\tmarker.db <- collapseLisToFrame(marker.lst)\t%>% \n",
       "\t\t\t`colnames<-`(c('celltype', 'gene')) %>% \n",
       "\t\t\tmutate(\n",
       "\t\t\t\tspecies = 'Human', \n",
       "\t\t\t\ttissue = NA, \n",
       "\t\t\t\tcancer = 'Normal', \n",
       "\t\t\t\tcondition = 'Normal', \n",
       "\t\t\t\tsubtype1 = NA, \n",
       "\t\t\t\tsubtype2 = NA, \n",
       "\t\t\t\tsubtype3 = NA, \n",
       "\t\t\t\tresource = NA, \n",
       "\t\t\t\tpmid = NA\n",
       "\t\t\t)\n",
       "\t}\n",
       "\tobj <- scCATCH::findmarkergene(data, species = match.arg(species), if_use_custom_marker = TRUE, marker = marker.db)\n",
       "\tpred.res <- scCATCH::findcelltype(obj)\n",
       "\tobj.tmp <- RenameIdents(obj.seu, split(pred.res@celltype$cell_type, pred.res@celltype$cluster) %>% unlist(.))\n",
       "\tobj.seu$scCATCH <- Idents(obj.tmp) %>% as.vector\n",
       "\treturn(obj.seu)\n",
       "}\n",
       "<environment: namespace:scAnnoX>\n",
       "\n",
       "$cellID\n",
       "function(obj.seu, marker.lst = NULL, gset.len = 5, sig.cut = 2, species = c('Hs', 'Mm'), tissue.type = NULL, ...) {\t\n",
       "   assay.names <- obj.seu@assays %>% names\n",
       "   if (!('data' %in% assay.names)) obj.seu@assays$data <- obj.seu@assays$RNA\n",
       "   obj.seu <- CelliD::RunMCA(obj.seu)\n",
       "\tif (is.null(marker.lst)) {\n",
       "\t\tpanglao <- system.file('data', 'PanglaoDB_markers_27_Mar_2020.tsv.gz', package = 'scAnnoX') %>% read_tsv\n",
       "\t\tpanglao.all <- panglao %>% filter(str_detect(species, match.arg(species)))\n",
       "\t\tif (!is.null(tissue.type)) panglao.all <- panglao.all %>% filter(organ == tissue.type)\n",
       "\t\tpanglao.all <- panglao.all %>%  \n",
       "\t\t\tgroup_by(`cell type`) %>%  \n",
       "\t\t\tsummarise(geneset = list(`official gene symbol`))\n",
       "\t\tall.gs <- setNames(panglao.all$geneset, panglao.all$`cell type`)\n",
       "\t\tall.gs <- all.gs[sapply(all.gs, length) >= gset.len]\n",
       "\t} else all.gs <- marker.lst[sapply(marker.lst, length) >= gset.len]\n",
       "   \n",
       "\tHGT.all.gs <- CelliD::RunCellHGT(obj.seu, pathways = all.gs, dims = 1 : 50)\n",
       "\tall.gs.pred <- rownames(HGT.all.gs)[apply(HGT.all.gs, 2, which.max)]\n",
       "\tall.gs.pred.signif <- ifelse(apply(HGT.all.gs, 2, max) > 2, yes = all.gs.pred, \"unassigned\")\n",
       "\tobj.seu$CellIDAnno <- all.gs.pred\n",
       "\tobj.seu$CellIDAnno.adj <- all.gs.pred.signif\n",
       "\treturn(obj.seu)\n",
       "}\n",
       "<environment: namespace:scAnnoX>\n",
       "\n",
       "$sc.type\n",
       "function(obj.seu, marker.lst = NULL, tissue.type = 'Immune system', ...) {\n",
       "\tsystem.file('modules', 'sc-type-master/R/auto_detect_tissue_type.R', package = 'scAnnoX') %>% source(.)\n",
       "\tsystem.file('modules', 'sc-type-master/R/gene_sets_prepare.R', package = 'scAnnoX') %>% source(.)\n",
       "\tsystem.file('modules', 'sc-type-master/R/sctype_score_.R', package = 'scAnnoX') %>% source(.)\n",
       "\t\n",
       "\tmarker.dbfile <- system.file('modules', 'sc-type-master/ScTypeDB_full.xlsx', package = 'scAnnoX')\n",
       "\tif (is.null(marker.lst)) {\n",
       "\t\tgs.list <- gene_sets_prepare(marker.dbfile, tissue.type)\n",
       "\t} else {\n",
       "\t\tgs.neg <- lapply(names(marker.lst), function(xx) character(0)) %>% `names<-`(names(marker.lst))\n",
       "\t\tgs.pos <- marker.lst\n",
       "\t\tgs.list <- list(gs_positive = gs.pos, gs_negative = gs.neg)\n",
       "\t}\n",
       "    sc.expr <- GetAssayData(obj.seu) %>% as.matrix\n",
       "    es.max <- sctype_score(scRNAseqData = sc.expr, gs = gs.list$gs_positive, gs2 = gs.list$gs_negative, ...)\n",
       "    pred.res <- apply(es.max, 2, which.max) %>% rownames(es.max)[.]\n",
       "\tobj.seu$scTypeAnno <- pred.res\n",
       "    return(obj.seu)\n",
       "}\n",
       "<environment: namespace:scAnnoX>\n",
       "\n",
       "$SCINA\n",
       "function(obj.seu, marker.lst, ...) {\n",
       "    pred.res <- SCINA::SCINA(\n",
       "\tobj.seu %>% GetAssayData(., slot = 'data') %>% as.matrix, \n",
       "\tmarker.lst, \n",
       "\tmax_iter = 1000, \n",
       "        convergence_n = 12, \n",
       "        convergence_rate = 0.999, \n",
       "\tsensitivity_cutoff = 0.9, \n",
       "\t...\n",
       "    )\n",
       "    obj.seu$SCINA <- pred.res$cell_labels\t\n",
       "    return(obj.seu)\n",
       "}\n",
       "<environment: namespace:scAnnoX>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "listToolMethods()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5091a46c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
